gossip/src/duplicate_shred_handler.rs

```diff
@@ -90,17 +97,43 @@ impl DuplicateShredHandler {
     pub fn new(
         blockstore: Arc<Blockstore>,
         leader_schedule_cache: Arc<LeaderScheduleCache>,
+        bank_forks: Arc<RwLock<BankForks>>,
     ) -> Self {
         Self {
             chunk_map: HashMap::new(),
             validator_pending_proof_map: HashMap::new(),
             last_root: 0,
+            last_root_cleaned: 0,
+            cached_on_slot: 0,
+            cached_staked_nodes: None,
+            cached_slots_in_epoch: 100,
             blockstore,
             leader_schedule_cache,
+            bank_forks,
             cleanup_count: CLEANUP_EVERY_N_LOOPS,
         }
     }
 
+    fn cache_root_info(&mut self) {
+        let last_root = self.blockstore.last_root();
+        if self.cached_staked_nodes.is_none() || last_root != self.last_root {
+            self.last_root = last_root;
+            // The stuff we need only changes per epoch, so we only need to cache
+            // once per epoch, and it's okay if we use old data in a few slots.
+            if self.cached_staked_nodes.is_some()
+                && last_root < self.cached_slots_in_epoch + self.cached_on_slot
+            {
+                return;
+            }
+            let bank_forks_result = self.bank_forks.read().unwrap();
```

Comment:
this is holding the lock on bank_forks unnecessarily for too long. instead:
```rust
let root_bank = self.bank_forks.read().unwrap().root_bank();
```

Reply:
done